### PR TITLE
Fix Floating IP idempotence (lack of)

### LINKED ISSUE #5: digital_ocean_floating_ip is not idempotent
_From @jackivanov on Oct 06, 2019 13:11_

##### SUMMARY
`digital_ocean_floating_ip` is not idempotent, this is important.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
digital_ocean_floating_ip

##### ANSIBLE VERSION

```paste below
ansible 2.8.3
  config file = /Users/XXX/XXX/ansible.cfg
  configured module search path = ['/Users/XXX/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/XXX/XXX/.env/lib/python3.7/site-packages/ansible
  executable location = /Users/XXX/XXX/.env/bin/ansible
  python version = 3.7.4 (default, Jul  9 2019, 18:13:23) [Clang 10.0.1 (clang-1001.0.46.4)]
```

##### STEPS TO REPRODUCE
1. Run the tasks below. Droplet gets created with a Floating IP assigned.
2. Run the tasks once again, the playbook fails on the second step.

```yaml
- name: "Creating a droplet..."
  digital_ocean_droplet:
    state: present
    name: "{{ name }}"
    oauth_token: "{{ token }}"
    size: "{{ size }}"
    region: "{{ egion }}"
    image: "{{ image }}"
    wait_timeout: 300
    unique_name: true
    ipv6: true
    ssh_keys: "{{ ssh_key }}"
  register: digital_ocean_droplet

- name: "Create a Floating IP"
  digital_ocean_floating_ip:
    state: present
    oauth_token: "{{ token }}"
    droplet_id: "{{ digital_ocean_droplet.data.droplet.id }}"
  register: digital_ocean_floating_ip
```

##### EXPECTED RESULTS
Ansible detects that a Floating IP is already assigned to the droplet and returns `success` with the data included.

##### ACTUAL RESULTS

```paste below
TASK [cloud-digitalocean : Create a Floating IP] ***********************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Error creating floating ip [422: Droplet is already assigned to another floating IP.]", "region": null}
```

_Copied from original issue: ansible/ansible#63175_

### PR DESCRIPTION
**Pull request by mamercad**
##### SUMMARY

Fixes #5.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

digital_ocean_floating_ip

##### ADDITIONAL INFORMATION

```paste below
❯ doctl compute floating-ip list
IP    Region    Droplet ID    Droplet Name
```

```
❯ ANSIBLE_STDOUT_CALLBACK=yaml ansible-playbook -i localhost -c local, test.yml
PLAY [test] ******************************************************************************************************************************************************

TASK [Creating a droplet...] *************************************************************************************************************************************
ok: [localhost]

TASK [Create a Floating IP] **************************************************************************************************************************************
changed: [localhost]

PLAY RECAP *******************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

```
❯ doctl compute floating-ip list
IP               Region    Droplet ID    Droplet Name
REDACTED    nyc3      211562865     hacktoberfest
```

```
❯ ANSIBLE_STDOUT_CALLBACK=yaml ansible-playbook -i localhost -c local, test.yml
PLAY [test] ******************************************************************************************************************************************************

TASK [Creating a droplet...] *************************************************************************************************************************************
ok: [localhost]

TASK [Create a Floating IP] **************************************************************************************************************************************
ok: [localhost]

PLAY RECAP *******************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

```

Note that `changed=0` is the behavior that we're after.

### CODE REVIEW

File: plugins/modules/digital_ocean_floating_ip.py [hunk 1]
```diff
@@ -270,6 +270,17 @@ def create_floating_ips(module, rest):
...
         payload["droplet_id"] = module.params['droplet_id']
 
+    # Get existing floating IPs
+    response = rest.get('floating_ips/')
+    status_code = response.status_code
+    json_data = response.json
+
+    # Exit unchanged if any of them are assigned to this Droplet already
+    if status_code == 200:
+        for floating_ip in json_data['floating_ips']:
```
Comment (mpontillo) [inline plugins/modules/digital_ocean_floating_ip.py:280]: Consider using `.get()` in this section of code, and checking to be sure the response JSON is not `None`.

**Review by mpontillo — DISMISSED** [review]

File: plugins/modules/digital_ocean_floating_ip.py [hunk 2]
```diff
@@ -287,6 +287,19 @@ def create_floating_ips(module, rest):
...
 
+    # Get existing floating IPs
+    response = rest.get('floating_ips/')
+    status_code = response.status_code
+    json_data = response.json
+
+    # Exit unchanged if any of them are assigned to this Droplet already
+    if status_code == 200:
+        for floating_ip in json_data.get('floating_ips', []):
+            if floating_ip.get('droplet', None) is not None:
```
Comment (mpontillo) [inline plugins/modules/digital_ocean_floating_ip.py:298]: Consider making `droplet` a variable here so you don't have to re-index the dictionary.

**Review by mpontillo — APPROVED** [review]

File: plugins/modules/digital_ocean_floating_ip.py (hunk 2 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_floating_ip.py]: Sounds good, this should be fixed now!

File: plugins/modules/digital_ocean_floating_ip.py [hunk 3]
```diff
@@ -287,6 +287,21 @@ def create_floating_ips(module, rest):
...
+
+    # Exit unchanged if any of them are assigned to this Droplet already
+    if status_code == 200:
+        for floating_ip in json_data.get('floating_ips', []):
+            droplet = floating_ip.get('droplet', None)
+            if droplet is not None:
+                droplet_id = droplet.get('id', None)
+                if droplet_id is not None:
+                    if droplet_id == module.params['droplet_id']:
+                        module.exit_json(changed=False, data={'floating_ip': floating_ip['ip']})
```
Comment (mpontillo) [inline plugins/modules/digital_ocean_floating_ip.py:303]: Consider `floating_ip.get('ip', None)` here.

**Review by mpontillo — DISMISSED** [review]

File: plugins/modules/digital_ocean_floating_ip.py (hunk 3 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_floating_ip.py]: Cool, I think this is in better shape now.

**Review by mpontillo — DISMISSED** [review]
lgtm!

**Comment by mamercad** [comment]
Sorry @mpontillo, this should be good now, one more look please when you have time.

File: plugins/modules/digital_ocean_floating_ip.py [hunk 4]
```diff
@@ -294,17 +294,19 @@ def create_floating_ips(module, rest):
...
-                        else:
-                            module.fail_json(changed=False, msg="Unexpected error querying floating ip")
+        floating_ips = json_data.get('floating_ips', [])
+        if floating_ips != []:
+            for floating_ip in floating_ips:
+                droplet = floating_ip.get('droplet', None)
+                if droplet is not None:
+                    droplet_id = droplet.get('id', None)
+                    if droplet_id is not None:
+                        if str(droplet_id) == module.params['droplet_id']:
```
Comment (mpontillo) [inline plugins/modules/digital_ocean_floating_ip.py:304]: Nice catch!

File: plugins/modules/digital_ocean_floating_ip.py [hunk 5]
```diff
@@ -294,17 +294,19 @@ def create_floating_ips(module, rest):
...
-                droplet_id = droplet.get('id', None)
-                if droplet_id is not None:
-                    if droplet_id == module.params['droplet_id']:
-                        ip = floating_ip.get('ip', None)
-                        if ip is not None:
-                            module.exit_json(changed=False, data={'floating_ip': ip})
-                        else:
-                            module.fail_json(changed=False, msg="Unexpected error querying floating ip")
+        floating_ips = json_data.get('floating_ips', [])
+        if floating_ips != []:
```
Comment (mpontillo) [inline plugins/modules/digital_ocean_floating_ip.py:298]: I think "this is fine", but how to check for an empty list is a subject of [immense debate](https://stackoverflow.com/questions/53513/how-do-i-check-if-a-list-is-empty) in the Python community. =)

You might consider doing:

```
if floating_ips:
```

This relies on boolean coercion rules, such as:
```
>>> bool([])
False
>>> bool(['127.0.0.1'])
True
```

Or you could do:
```
if len(floating_ips) != 0:
```

**Review by mpontillo — DISMISSED** [review]

File: plugins/modules/digital_ocean_floating_ip.py (hunk 5 above)
Comment (mamercad) [reply plugins/modules/digital_ocean_floating_ip.py]: I went with the last suggestion, thank you!

**Review by mpontillo — APPROVED** [review]
Thanks for the fixes!

**Comment by mamercad** [comment]
> Thanks for the fixes!

Thanks for the reviews!

**Pull request merged**